### PR TITLE
ics20-tao-v2-support

### PR DESCRIPTION
**Pull request by sangier**
This PR introduces IBC v2 TAO support for the ics20 v1 and v2 trying to capture all the internal discussions. 

Closes https://github.com/cosmos/ibc/issues/1141

### CODE REVIEW

File: spec/app/ics-020-fungible-token-transfer/v1/README.md [hunk 1]
```diff
@@ -240,130 +146,142 @@ In plain English, between chains `A` and `B`:
...
 ```
 
 `onAcknowledgePacket` is called by the routing module when a packet sent by this module has been acknowledged.
 
 ```typescript
 function onAcknowledgePacket(
-  packet: Packet,
-  acknowledgement: bytes) {
+  sourceChannelId: bytes,
+  destChannelId: bytes, // Can be nullable
```
Comment (AdityaSripal) [inline spec/app/ics-020-fungible-token-transfer/v1/README.md:232]: Why would this be nullable?

File: spec/app/ics-020-fungible-token-transfer/README.md [hunk 2]
```diff
@@ -351,75 +248,47 @@ In plain English, between chains `A` and `B`:
...
-  channel = provableStore.get(channelPath(sourcePort, sourceChannel))
-  // getAppVersion returns the transfer version that is embedded in the channel version
-  // as the channel version may contain additional app or middleware version(s)
-  transferVersion = getAppVersion(channel.version)
+  transferVersion = payload.version
   if transferVersion == "ics20-1" {
-    abortTransactionUnless(len(tokens) == 1)
-    token = tokens[0]
+    abortTransactionUnless(len(appData.tokens) == 1)
+    token = appData.tokens[0]
```
Comment (AdityaSripal) [inline spec/app/ics-020-fungible-token-transfer/README.md:282]:
```suggestion
```

File: spec/app/ics-020-fungible-token-transfer/README.md [hunk 3]
```diff
@@ -351,75 +248,47 @@ In plain English, between chains `A` and `B`:
...
+      bank.BurnCoins(appData.sender, onChainDenom, token.amount)
     }
   }
 
-  var dataBytes bytes
-  channel = provableStore.get(channelPath(sourcePort, sourceChannel))
-  // getAppVersion returns the transfer version that is embedded in the channel version
-  // as the channel version may contain additional app or middleware version(s)
-  transferVersion = getAppVersion(channel.version)
+  transferVersion = payload.version
```
Comment (AdityaSripal) [inline spec/app/ics-020-fungible-token-transfer/README.md:279]: We can move this block before the escrowing logic

File: spec/app/ics-020-fungible-token-transfer/README.md [hunk 4]
```diff
@@ -428,18 +297,20 @@ function sendFungibleTokens(
...
 ```typescript
-function onRecvPacket(packet: Packet) {
-  channel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
-  // getAppVersion returns the transfer version that is embedded in the channel version
-  // as the channel version may contain additional app or middleware version(s)
-  transferVersion = getAppVersion(channel.version)
+function onRecvPacket(
+  destChannelId: bytes,
+  sourceChannelId: bytes,
+  sequence: bigEndianUint64,
```
Comment (AdityaSripal) [inline spec/app/ics-020-fungible-token-transfer/README.md:303]:
```suggestion
  sequence: uint64, 
```

File: spec/app/ics-020-fungible-token-transfer/README.md [hunk 5]
```diff
@@ -529,114 +400,131 @@ function onRecvPacket(packet: Packet) {
...
-  prevPacket = privateStore.get(packetForwardPath(packet.sourcePort, packet.sourceChannel, packet.sequence))
+  prevPacketSeq = privateStore.get(packetForwardPath(sourceChannelId, sequence))
 
-  if prevPacket != nil {
+  if prevPacketSeq != nil {
     if acknowledgement.success {
       FungibleTokenPacketAcknowledgement ack = FungibleTokenPacketAcknowledgement{true, "forwarded packet succeeded"}
       handler.writeAcknowledgement(
-        prevPacket,
+        destChannelId,
```
Comment (AdityaSripal) [inline spec/app/ics-020-fungible-token-transfer/README.md:479]: We aren't using the destChannelId here. We need the dest channel of the **previous packet**

If its not available here then we need to store it along with the previous sequence.

File: spec/app/ics-020-fungible-token-transfer/README.md [hunk 6]
```diff
@@ -529,114 +400,131 @@ function onRecvPacket(packet: Packet) {
...
       // the forwarded packet has failed, thus the funds have been refunded to the forwarding address.
       // we must revert the changes that came from successfully receiving the tokens on our chain
       // before propogating the error acknowledgement back to original sender chain
-      revertInFlightChanges(packet, prevPacket)
+      revertInFlightChanges(destChannelId, payload)
       // write error acknowledgement
       FungibleTokenPacketAcknowledgement ack = FungibleTokenPacketAcknowledgement{false, "forwarded packet failed"}
       handler.writeAcknowledgement(
-        prevPacket,
+        destChannelId,
```
Comment (AdityaSripal) [inline spec/app/ics-020-fungible-token-transfer/README.md:491]: ditto

File: spec/app/ics-020-fungible-token-transfer/README.md [hunk 7]
```diff
@@ -529,114 +400,131 @@ function onRecvPacket(packet: Packet) {
...
-  if prevPacket != nil {
+  if prevPacketSeq != nil {
     // the forwarded packet has failed, thus the funds have been refunded to the forwarding address.
     // we must revert the changes that came from successfully receiving the tokens on our chain
     // before propogating the error acknowledgement back to original sender chain
-    revertInFlightChanges(packet, prevPacket)
+    revertInFlightChanges(destChannelId, payload)
     // write error acknowledgement
     FungibleTokenPacketAcknowledgement ack = FungibleTokenPacketAcknowledgement{false, "forwarded packet timed out"}
     handler.writeAcknowledgement(
```
Comment (AdityaSripal) [inline spec/app/ics-020-fungible-token-transfer/README.md:530]: This needs to change to not have `prevPacket` in argument

File: spec/app/ics-020-fungible-token-transfer/README.md [hunk 8]
```diff
@@ -716,19 +604,22 @@ function refundTokens(packet: Packet) {
 // If an error occurs further down the line, the state changes
 // on this chain must be reverted before sending back the error acknowledgement
 // to ensure atomic packet forwarding
-function revertInFlightChanges(sentPacket: Packet, receivedPacket: Packet) {
-  forwardingAddress = channelForwardingAddress[receivedPacket.destChannel]
-  reverseEscrow = channelEscrowAddresses[receivedPacket.destChannel]
+function revertInFlightChanges(
```
Comment (AdityaSripal) [inline spec/app/ics-020-fungible-token-transfer/README.md:607]: This function was incorrectly converted. Note that it originally took in **two** different packets and now you only have a single destChannelId and payload. You need to still take in the forwarded packet info and the original received packet info

**Review by AdityaSripal — CHANGES_REQUESTED** [review]
Nice start! The ics20-v1 spec looks correct, but there are some mistakes on the ics20-v2 side

File: spec/app/ics-020-fungible-token-transfer/v1/README.md (hunk 1 above)
Comment (sangier) [reply spec/app/ics-020-fungible-token-transfer/v1/README.md:232]: Yeah, here what I wanted to express is that we call the `onAckPacket` with the parameters specified in the [ics04](https://github.com/cosmos/ibc/blob/84341283c480bcfbd982a58b2d0626255eeaadf1/spec/core/v2/ics-004-channel-and-packet-semantics/README.md#:~:text=OnAcknowledgePacket(packet.channelSourceId%2Cpacket.channelDestId%2Cpacket.sequence%2Cpayload%2Cacknowledgement%2C%20relayer)), but, actually, we don't use all of them. Indeed, while all parameters are necessary for [ics20v2](https://github.com/cosmos/ibc/blob/99eaf71eeae605c521261d54765858b1984de28a/spec/app/ics-020-fungible-token-transfer/README.md#:~:text=function%20onAcknowledgePacket(%0A%20%20sourceChannelId%3A%20bytes%2C%0A%20%20destChannelId%3A%20bytes%2C%20%0A%20%20sequence%3A%20bigEndianUint64%2C%0A%20%20payload%3A%20Payload%2C%20%0A%20%20acknowledgement%3A%20bytes%0A%20%20)%3A%20bool%20%7B) `onAckPacket` logic, `destChannelId` and `sequence` are not used in the ics20v1 `onAckPacket`. 

So the question is how do we want to handle this difference in the interfaces? 
1. Do we want to have a standard way to call applications on the core side? 
2. Do we let the core understand how to call the specific app callback retrieving from the router the callback signature? 

Option 2 should provide more flexibility and seems ok. What do you think?

File: spec/app/ics-020-fungible-token-transfer/README.md [hunk 9]
```diff
@@ -428,18 +297,20 @@ function sendFungibleTokens(
...
+  ): (bytes, bool) {
+  transferVersion = payload.version
   var tokens []Token
   var sender string
   var receiver string // address to send tokens to on this chain
   var finalReceiver string // final intended address in forwarding case
 
   if transferVersion == "ics20-1" {
-     FungibleTokenPacketData data = json.unmarshal(packet.data)
+     FungibleTokenPacketData data = payload.encoding.unmarshal(payload.appData)
```
[the remark this answers is not in the document]
Comment (AdityaSripal) [reply spec/app/ics-020-fungible-token-transfer/README.md]: payload.encoding is just a string

File: spec/app/ics-020-fungible-token-transfer/README.md [hunk 10]
```diff
@@ -529,123 +418,143 @@ function onRecvPacket(packet: Packet) {
...
-    refundTokens(packet)
+    refundTokens(sourceChannelId, payload)
   }
 
   // check if the packet that was sent is from a previously forwarded packet
-  prevPacket = privateStore.get(packetForwardPath(packet.sourcePort, packet.sourceChannel, packet.sequence))
+  prevPacketSeq,prevPacketDestChannelId, prevPacketDestPort = privateStore.get(packetForwardPath(sourceChannelId, sequence))
 
-  if prevPacket != nil {
+  if prevPacketSeq != nil {
```
Comment (AdityaSripal) [inline spec/app/ics-020-fungible-token-transfer/README.md:494]:
```suggestion
  if prevPacketSeq != 0 {
```

File: spec/app/ics-020-fungible-token-transfer/README.md [hunk 11]
```diff
@@ -529,123 +418,143 @@ function onRecvPacket(packet: Packet) {
...
     // the forwarded packet has failed, thus the funds have been refunded to the forwarding address.
     // we must revert the changes that came from successfully receiving the tokens on our chain
     // before propogating the error acknowledgement back to original sender chain
-    revertInFlightChanges(packet, prevPacket)
+    revertInFlightChanges(destChannelId, payload, prevPacketDestChannelId,prevPacketDestPort)
     // write error acknowledgement
     FungibleTokenPacketAcknowledgement ack = FungibleTokenPacketAcknowledgement{false, "forwarded packet timed out"}
     handler.writeAcknowledgement(
+      prevPacketDestChannelId
       prevPacket,
```
Comment (AdityaSripal) [inline spec/app/ics-020-fungible-token-transfer/README.md:552]:
```suggestion
      prevPacketSeq,
```

**Review by AdityaSripal — APPROVED** [review]
Excellent work! Everything looks good now, will commit the small fixes and then merge
